lib/domain/usecase/contacts/phonebook_contact_interactor_v2.dart

```diff
@@ -0,0 +1,288 @@
+
+    final contactIdToHashMap = {
+      for (final contact in contacts) contact.id: contact,
+    };
+
+    final chunks = contactIdToHashMap.values.slices(lookupChunkSize);
+
+    for (final chunkContacts in chunks) {
+      final Map<String, List<String>> hashToContactIdMappings = {};
+      for (final contact in chunkContacts) {
+        final updatedContact = processContacts(
+          chunkContacts,
+          hashDetails,
+          hashToContactIdMappings,
+          contactIdToHashMap,
+        );
+
+        if (updatedContact == null) continue;
+
+        contactIdToHashMap[contact.id] = updatedContact;
+      }
+
+      final response = await _lookupRepository.lookupListMxid(
+        LookupListMxidRequest(
+          addresses:
+              hashToContactIdMappings.values.expand((hash) => hash).toSet(),
+          algorithm: hashDetails.algorithms?.firstOrNull,
+          pepper: hashDetails.lookupPepper,
+        ),
+      );
+
+      final totalChunks = chunks.length;
+
+      yield await _handleLookupMappings(
+        progress: lookupChunkSize ~/ totalChunks,
+        chunkSize: lookupChunkSize,
+        totalChunks: totalChunks,
+        mappings: response.mappings ?? {},
+        hashToContactIdMappings: hashToContactIdMappings,
+        chunkContacts: chunkContacts,
+      );
+      _handleLookupThirdPartyId(
+        thirdPartyMappings: response.thirdPartyMappings ?? {},
+      );
+    }
+  }
+
+  Map<String, List<String>> calculateHashesForPhoneNumbers(
+    Set<PhoneNumber> phoneNumbers,
+    HashDetailsResponse hashDetails,
+  ) {
+    final Map<String, List<String>> phoneToHashMap = {};
+    for (final phoneNumber in phoneNumbers) {
+      final hashes = phoneNumber.calculateHashUsingAllPeppers(
+        hashDetails: hashDetails,
+      );
+      phoneToHashMap[phoneNumber.number] = hashes;
+      phoneNumber.setThirdPartyIdToHashMap(hashes);
+    }
+    return phoneToHashMap;
+  }
+
+  Map<String, List<String>> calculateHashesForEmails(
+    Set<Email> emails,
+    HashDetailsResponse hashDetails,
+  ) {
+    final Map<String, List<String>> emailToHashMap = {};
+    for (final email in emails) {
+      final hashes = email.calculateHashUsingAllPeppers(
+        hashDetails: hashDetails,
+      );
+      emailToHashMap[email.address] = hashes;
+      email.setThirdPartyIdToHashMap(hashes);
+    }
+    return emailToHashMap;
+  }
+
+  Contact updateContactWithHashes(
+    Contact contact,
+    Map<String, List<String>> phoneToHashMap,
+    Map<String, List<String>> emailToHashMap,
+  ) {
+    final updatedPhoneNumbers = <PhoneNumber>{};
+    final updatedEmails = <Email>{};
+
+    for (final phoneNumber in contact.phoneNumbers!) {
+      updatedPhoneNumbers.add(phoneNumber);
+    }
+
+    for (final email in contact.emails!) {
+      updatedEmails.add(email);
+    }
+
+    return contact.copyWith(
+      phoneNumbers: updatedPhoneNumbers,
+      emails: updatedEmails,
+    );
+  }
+
+  Contact? processContacts(
```

Comment:
specific in what kind of process? 
```suggestion
  Contact? calculateHashForContact(
```